libraries/Pliant/Charts/PreComputedSet.cs

```diff
@@ -1,15 +1,29 @@
 ﻿using Pliant.Collections;
 using Pliant.Grammars;
 using System.Collections.Generic;
+using System.Linq;
 
 namespace Pliant.Charts
 {
     internal class PreComputedSet
     {
         private readonly UniqueList<StateFrame> _frames;
+        private StateFrame[] _cachedFrames;
 
         public IReadOnlyList<StateFrame> Frames { get { return _frames; } }
 
+        public StateFrame[] FramesPerf
```

Comment:
I don't see any reason this couldn't be the default way that the frames are handled. May have to write a ArrayWrapper&lt;T&gt; class to handle the IReadOnlyList&lt;T&gth; implementation. Basically when the Frames list is first accessed, cache the frames to an array. Currently you are using ToArray() so that may change to new ArrayWrapper&lt;T&gt;(_frames.ToArray())